sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/AggUtils.scala

```diff
@@ -144,11 +145,16 @@ object AggUtils {
     // [COUNT(DISTINCT foo), MAX(DISTINCT foo)], but [COUNT(DISTINCT bar), COUNT(DISTINCT foo)] is
     // disallowed because those two distinct aggregates have different column expressions.
     val distinctExpressions = functionsWithDistinct.head.aggregateFunction.children
-    val namedDistinctExpressions = distinctExpressions.map {
-      case ne: NamedExpression => ne
-      case other => Alias(other, other.toString)()
+    val normalizedNamedDistinctExpressions = distinctExpressions.map { e =>
+      // Ideally this should be done in `NormalizeFloatingNumbers`, but we do it here because
+      // `groupingExpressions` is not extracted during logical phase.
+      NormalizeFloatingNumbers.normalize(e) match {
+        case ne: NamedExpression => ne
+        case other => Alias(other, other.toString)()
```

Comment:
This works correctly, @viirya . As an architectural perspective, I'm wondering if we can collect these `NormalizeFloatingNumbers.normalize` application in `SparkStrategies` instead of putting `AggUtils`. Actually, in `SparkStragegies`, we have the same logic for `normalizedGroupingExpressions` and missed this conversion at `functionsWithDistinct`. IMO, it would be easier if we keep this converting in one place, `SparkStragegies`, in order to avoid missing corner cases like this.

Reply:
Generally I agree. However, we need to use `distinctExpressions` for the `normalizedNamedDistinctExpressions`. 

If we want to put this `NormalizeFloatingNumbers` invocation in `SparkStrategies`, we need to move `distinctExpressions` from `AggUtils.planAggregateWithOneDistinct` to `SparkStrategies` too. 

It would look like, in `SparkStrategies`:

```scala
if (functionsWithDistinct.isEmpty) {
  AggUtils.planAggregateWithoutDistinct(
    normalizedGroupingExpressions,
    aggregateExpressions,
    resultExpressions,
    planLater(child))
} else {
  // functionsWithDistinct is guaranteed to be non-empty. Even though it may contain more than one
  // DISTINCT aggregate function, all of those functions will have the same column expressions.
  // For example, it would be valid for functionsWithDistinct to be
  // [COUNT(DISTINCT foo), MAX(DISTINCT foo)], but [COUNT(DISTINCT bar), COUNT(DISTINCT foo)] is
  // disallowed because those two distinct aggregates have different column expressions.
  val distinctExpressions = functionsWithDistinct.head.aggregateFunction.children
  val normalizedNamedDistinctExpressions = distinctExpressions.map { e =>
      // Ideally this should be done in `NormalizeFloatingNumbers`, but we do it here because
      // `groupingExpressions` is not extracted during logical phase.
      NormalizeFloatingNumbers.normalize(e) match {
        case ne: NamedExpression => ne
        case other => Alias(other, other.toString)()
      }
    }
  AggUtils.planAggregateWithOneDistinct(
    normalizedGroupingExpressions,
    functionsWithDistinct,
    functionsWithoutDistinct,
    distinctExpressions,
    normalizedNamedDistinctExpressions,
    resultExpressions,
    planLater(child))
}
```

This leaks more details from `AggUtil` in `SparkStrategies`. Looks not pretty good.

Reply:
Rather, we cannot move the existing `NormalizeFloatingNumbers` logic (int `SparkStrategies`) into AggUtils?

Reply:
We could have `NormalizeFloatingNumbers.normalize` in `AggUtils`, but we need to have it both in `AggUtils.planAggregateWithoutDistinct` and `AggUtils.planAggregateWithOneDistinct`.

Reply:
If moving into `AggUtils` sounds better for you all, I can move it. WDYT?

Reply:
I left a comment for the same thing, and just resolved it by myself. I don't have a strong opinion on this.

Reply:
If we broaden the scope, `SparkStrategies` already is looking at the detail of `functionsWithDistinct` like the following.
```scala
        val (functionsWithDistinct, functionsWithoutDistinct) =
          aggregateExpressions.partition(_.isDistinct)
        if (functionsWithDistinct.map(_.aggregateFunction.children.toSet).distinct.length > 1) {
          // This is a sanity check. We should not reach here when we have multiple distinct
          // column sets. Our `RewriteDistinctAggregates` should take care this case.
          sys.error("You hit a query analyzer bug. Please report your query to " +
              "Spark user mailing list.")
        }
```

And the very next line is the same logic block for `groupingExpression`.
```scala
        // Ideally this should be done in `NormalizeFloatingNumbers`, but we do it here because
        // `groupingExpressions` is not extracted during logical phase.
        val normalizedGroupingExpressions = groupingExpressions.map { e =>
          NormalizeFloatingNumbers.normalize(e) match {
            case n: NamedExpression => n
            case other => Alias(other, e.name)(exprId = e.exprId)
          }
        }
```

Given the above, I guess what you concerned is only one line source code, `val distinctExpressions = functionsWithDistinct.head.aggregateFunction.children`.

And, the comment is about the definition of `functionsWithDistinct` which is generated from the above. The comment is not a detail hidden from `SparkStrategies`. For me, it seems to be an assumption given by `SparkStrategies`.

Reply:
It looks weird if we do normalization in both `SparkStrategies` and `AggUtils`, better to put them in one place.

Reply:
Moved to `SparkStrategies`.